repl/src/main/scala/com/cloudera/livy/repl/Session.scala

```diff
@@ -82,21 +84,46 @@ class Session(interpreter: Interpreter, stateChangedCallback: SessionState => Un
   def execute(code: String): Int = {
     val statementId = newStatementId.getAndIncrement()
     _statements(statementId) = new Statement(statementId, StatementState.Waiting, null)
+
     Future {
-      _statements(statementId) = new Statement(statementId, StatementState.Running, null)
+      _statements(statementId).state.compareAndSet(StatementState.Waiting, StatementState.Running)
 
-      _statements(statementId) =
-        new Statement(statementId, StatementState.Available, executeCode(statementId, code))
+      setJobGroup(statementId)
+      val executeResult = if (_statements(statementId).state.get() != StatementState.Cancelling) {
```

Comment:
I think it's a bug we should really fix. If we hit this in a simple unit test, users will definitely hit it in production.